tests/image/imageengine.js

```diff
@@ -168,7 +168,111 @@ describe( 'ImageEngine', () => {
 
 				editor.setData( '<figure class="image"><img src="foo.jpg" /><img src="bar.jpg" />abc</figure>' );
 
-				expect( getModelData( document, { withoutSelection: true } ) ).to.equal( '<image src="foo.jpg">abc</image>' );
+				// The foo.jpg image is properly converted using figure converter. The other image was tried to
+				// be added as a child of foo.jpg and then was autohoisted.
+				expect( getModelData( document, { withoutSelection: true } ) )
+					.to.equal( '<image src="bar.jpg"></image><image src="foo.jpg">abc</image>' );
+			} );
+
+			describe( 'should autohoist images', () => {
+				beforeEach( () => {
+					document.schema.registerItem( 'div', '$block' );
+
+					buildModelConverter().for( editor.data.modelToView, editor.editing.modelToView ).fromElement( 'div' ).toElement( 'div' );
+					buildViewConverter().for( editor.data.viewToModel ).fromElement( 'div' ).toElement( 'div' );
+				} );
+
+				it( 'image between non-hoisted elements', () => {
+					editor.setData( '<div>foo<img src="foo.jpg" alt="foo" />bar</div>' );
+
+					expect( getModelData( document, { withoutSelection: true } ) ).to.equal(
+						'<div>foo</div>' +
+						'<image alt="foo" src="foo.jpg"></image>' +
+						'<div>bar</div>'
+					);
+				} );
+
+				it( 'multiple images', () => {
```

Comment:
Uh... I looked for that test being afraid that you haven't thought about it :D 🥇